spec/services/create_reservation_spec.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+describe CreateReservation do
+  let(:user) { User.create(user_type: :guest) }
+  let(:event_date_id) { unique_event_date_id }
+
+  let(:pantry_finder_api) { instance_double(PantryFinderApi) }
+  let(:capacity) { 100 }
+
+  let(:service) do
+    described_class.new(user_id: user.id, event_date_id: event_date_id)
+  end
+  let(:service_call) { service.call }
+
+  before do
+    allow(PantryFinderApi).to receive(:new).and_return(pantry_finder_api)
+    allow(pantry_finder_api).to receive(:event_date)
+      .with(event_date_id).and_return(capacity: capacity)
+  end
+
+  it 'creates a reservation' do
+    expect { service_call }.to change(Reservation, :count).by(1)
+    reservation = service_call.reservation
+
+    expect(service_call).to be_success
+    expect(reservation).to be_persisted
+    expect(reservation.user_id).to eq(user.id)
+    expect(reservation.event_date_id).to eq(event_date_id)
+  end
+
+  context 'when user already has a reservation for this event' do
+    before do
+      user.reservations.create!(event_date_id: event_date_id)
+    end
+
+    it 'does not create a reservation' do
+      expect { service_call }.not_to change(Reservation, :count)
+
+      expect(service_call).not_to be_success
+      expect(service_call.errors.full_messages).to eq(
+        ['User has already registered for this event']
+      )
+    end
+  end
+
+  context 'when event is at capacity' do
+    let(:capacity) { 0 }
+
+    it 'does not create a reservation' do
+      expect { service_call }.not_to change(Reservation, :count)
+
+      expect(service_call).not_to be_success
+      expect(service_call.errors.full_messages).to eq(
+        ['Event date is at capacity']
+      )
+    end
+  end
+
+  context 'when reservation fails to save' do
+    before do
+      allow(service.reservation).to receive(:save).and_return(false)
+    end
+
+    it 'is not successful' do
+      expect { service_call }.not_to change(Reservation, :count)
+
+      expect(service_call).not_to be_success
+    end
+  end
+
+  context 'when a faraday error occurs' do
+    before do
+      allow(pantry_finder_api).to receive(:event_date)
+        .with(event_date_id).and_raise(Faraday::Error.new('fail'))
+    end
+
+    it 'does not create a reservation' do
+      expect { service_call }.not_to change(Reservation, :count)
+
+      expect(service_call).not_to be_success
+      expect(service_call.errors.full_messages).to eq(
+        ['Pantry finder api fail']
+      )
+    end
+  end
+
+  def unique_event_date_id
```

Comment:
I'm not sure exactly what this does, but to me it seems like it is making the next unique event id? E.g. if the last event_date we entered was id 9654 it would then serve up 9655. I'm not sure what this is used for, but technically the next event_date_id should come from PantryTrak->FreshTrak and Freshtrak will never generate an event_date.

Reply:
Yup. This method does exactly what you think. It returns an integer 1 higher that the largest event date in the system. In production, these values will come from PantryTrak. This method is only used here for testing purposes. It's a mock value to prevent any dependencies on running other systems during unit tests. Neither this file or this method will ever be used in production.